[e2e-framework] e2ecluster/e2ehelpers: add cURL helper function

tklauser opened the pull request:
This allows to construct cURL commands to be executed inside pods and is
a more flexible replacement for the existing CurlFail, CurlFailNoStats,
CurlWithHTTPCode and CurlWithRetries helpers.

twpayne commented on pkg/e2ecluster/e2ehelpers/curl.go:
```diff
@@ -0,0 +1,105 @@
+
+const (
+	defaultConnectTimeout = 5  // seconds
+	defaultMaxTime        = 20 // seconds
```
nit: can these be `time.Duration`s?

tklauser replied:
Good point, I've changed the `CurlOpts` members and the respective `With*` methods as well.

twpayne commented on pkg/e2ecluster/e2ehelpers/curl.go:
```diff
@@ -0,0 +1,105 @@
+		cmd = append(cmd, "--fail")
+	}
+	if o.ResultFormat != "" {
+		cmd = append(cmd, "-w", fmt.Sprintf("%q", o.ResultFormat))
```
I don't think quotes are needed here as the argument is passed unchanged as an element of a `[]string` rather than as part of a larger string.

tklauser replied:
I think they are still needed at least in case of `CurlResultFormatStats` because that one contains spaces.

twpayne commented on pkg/e2ecluster/e2ehelpers/curl.go:
```diff
@@ -0,0 +1,105 @@
+		cmd = append(cmd, "--output", o.Output)
+	}
+	if o.Retries > 0 {
+		cmd = append(cmd, "--retry", fmt.Sprintf("%d", o.Retries))
```
tiny nit: `strconv.Itoa` avoids the overhead of parsing format strings. Similarly below.

tklauser replied:
Ack for this case. ~For the ones below I had to switch to `fmt.Sprintf("%.f")` because the timeouts are now `time.Duration`s.~ Actually, `strconv.FormatFloat` will work better than `fmt.Sprintf` in that case.

twpayne commented on pkg/e2ecluster/e2ehelpers/curl.go:
```diff
@@ -0,0 +1,105 @@
+	return o
+}
+
+func Curl(url string, opts ...CurlOption) []string {
```
Suggestion: rename function to `CurlCommandAndArgs` or similar, `Curl` on its own might be understood to execute the command.

gandro commented on pkg/e2ecluster/e2ehelpers/curl.go:
```diff
@@ -0,0 +1,107 @@
+	CurlResultFormatHTTPCode CurlResultFormat = "%%{http_code}"
+)
+
+type CurlOpts struct {
```
Overall style question: Does this struct need to be public? I noticed it is public in other helpers too, but I don't think it needs to be.

tklauser replied:
Good find, I don't think it needs to be public. I'll change this and add another commit to change existing `*Opts` structs.

gandro commented on pkg/e2ecluster/e2ehelpers/curl.go:
```diff
@@ -0,0 +1,107 @@
+	CurlResultFormatHTTPCode CurlResultFormat = "%%{http_code}"
+)
+
+type opts struct {
```
nit: Rename this to `curlOpts`, given that the e2ehelper package will get other helpers with an opts type too.

tklauser committed:
e2ecluster/e2ehelpers: add cURL helper function

This allows to construct cURL commands to be executed inside pods and is
a more flexible replacement for the existing CurlFail, CurlFailNoStats,
CurlWithHTTPCode and CurlWithRetries helpers.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
e2ecluster/ciliuminstall: unexport options struct

This doesn't need to be public.

Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

willfindlay reviewed (approved):
This looks amazing!
